packages/core/src/workflow/context.ts

```diff
@@ -1,111 +1,510 @@
+    if (this.#iterator === null) {
+      this.#iterator = this.#createStreamEvents();
     }
-    if (step === undefined) {
-      for (const queue of this.#queues.values()) {
-        queue.push(message);
+    return this.#iterator;
+  }
+
+  [Symbol.asyncIterator](): AsyncIterableIterator<WorkflowEvent> {
+    return this.#iteratorSingleton;
+  }
+
+  snapshot(): ArrayBuffer {
+    const state = {
+      startEvent: this.#startEvent,
+      queue: this.#queue,
+      pendingInputQueue: this.#pendingInputQueue,
+      data: this.#data,
+      timeout: this.#timeout,
+      verbose: this.#verbose,
+      resolved: this.#resolved,
+      rejected: this.#rejected,
+    };
+
+    const jsonString = JSON.stringify(state, (_, value) => {
+      // If value is an instance of a class, serialize only its properties
+      if (value instanceof WorkflowEvent) {
+        return { data: value.data, constructor: value.constructor.name };
       }
-    } else {
-      if (!this.#workflow.hasStep(step)) {
-        throw new Error(`Step ${step} does not exist`);
+      // value is Subtype of WorkflowEvent
+      if (
+        typeof value === "object" &&
+        value !== null &&
+        value?.prototype instanceof WorkflowEvent
+      ) {
+        return { constructor: value.prototype.constructor.name };
       }
+      return value;
+    });
 
-      if (!this.#queues.has(step)) {
-        this.#queues.set(step, []);
-      }
-      this.#queues.get(step)!.push(message);
-    }
+    return new TextEncoder().encode(jsonString).buffer;
   }
 
-  getNextEvent(step: StepFunction): WorkflowEvent | undefined {
-    const queue = this.#queues.get(step);
-    if (queue && queue.length > 0) {
-      return queue.shift();
-    }
-    return undefined;
+  #sendEvent = (event: WorkflowEvent): void => {
+    this.#queue.push({
+      type: "event",
+      event,
+    });
+  };
+
+  #requireEvent = async (
+    event: typeof WorkflowEvent<any>,
+  ): Promise<WorkflowEvent> => {
+    const requestId = randomUUID();
+    this.#queue.push({
+      type: "requestEvent",
+      id: requestId,
+      requestEvent: event,
+    });
+    return new Promise((resolve) => {
+      const handler = (event: InstanceType<GlobalEvent>) => {
+        if (event instanceof CustomEvent) {
+          const { id } = event.detail;
+          if (requestId === id) {
+            this.#queueEventTarget.removeEventListener("update", handler);
+            resolve(event.detail.event);
+          }
+        }
+      };
+      this.#queueEventTarget.addEventListener("update", handler);
+    });
+  };
+
+  #pendingInputQueue: WorkflowEvent[] = [];
+
+  /**
+   * Stream events from the start event
+   *
+   * Note that this function will stop once there's no more future events,
+   *  if you want stop immediately once reach a StopEvent, you should handle it in the other side.
+   * @private
+   */
+  #createStreamEvents(): AsyncIterableIterator<WorkflowEvent> {
```

Comment:
1. can we reduce the complexity of this function?
2. how can we implement to limit the number of promises per step - in python, that's: `@step(num_workers=4)`